client/pd_service_discovery.go

```diff
@@ -837,42 +950,106 @@ func (c *pdServiceDiscovery) updateURLs(members []*pdpb.Member) {
 	log.Info("[pd] update member urls", zap.Strings("old-urls", oldURLs), zap.Strings("new-urls", urls))
 }
 
-func (c *pdServiceDiscovery) switchLeader(addrs []string) error {
+func (c *pdServiceDiscovery) switchLeader(addrs []string) (bool, error) {
 	// FIXME: How to safely compare leader urls? For now, only allows one client url.
 	addr := addrs[0]
-	oldLeader := c.getLeaderAddr()
-	if addr == oldLeader {
-		return nil
+	oldLeader := c.getLeaderServiceClient()
+	if addr == oldLeader.GetAddress() && oldLeader.GetClientConn() != nil {
+		return false, nil
 	}
 
-	if _, err := c.GetOrCreateGRPCConn(addr); err != nil {
-		log.Warn("[pd] failed to connect leader", zap.String("leader", addr), errs.ZapError(err))
+	newConn, err := c.GetOrCreateGRPCConn(addr)
+	// If gRPC connect is created successfully or leader is new, still saves.
+	if addr != oldLeader.GetAddress() || newConn != nil {
+		// Set PD leader and Global TSO Allocator (which is also the PD leader)
+		leaderClient := newPDServiceClient(addr, addr, newConn, true)
+		c.leader.Store(leaderClient)
```

Comment:
What does the c.leader actually save?

Reply:
Saves `ServiceClient` including address and connection